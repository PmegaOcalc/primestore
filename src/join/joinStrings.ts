export function joinStrings<T extends string[], D extends string = "_">(
  array: T,
  divider: D = "_" as D
): Join<T, D>[] {
  const result = [];
  for (let i = 0; i < array.length; i++) {
    for (let j = i + 1; j < array.length + 1; j++) {
      result.push(array.slice(i, j).join(divider));
    }
  }
  return result as Join<T, D>[];
}

export type Join<
  T extends string[],
  divider extends string,
  prefix extends string = never
> = T extends [infer first extends string, ...infer rest extends string[]]
  ?
      | first
      | (rest extends never
          ? never
          :
              | Join<rest, divider, prefix>
              | Join<rest, divider, first>
              | (prefix extends string
                  ?
                      | `${prefix}${divider}${first}`
                      | Join<rest, divider, `${prefix}${divider}${first}`>
                  : never))
  : never;
