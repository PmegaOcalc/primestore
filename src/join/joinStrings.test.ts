import { joinStrings } from "./joinStrings";
import { Statuses } from "./test.types";

describe("joinStrings function", () => {
  test("to be defined", () => {
    expect(joinStrings).toBeDefined();
  });

  test("renders all possible strings correctly", () => {
    const samples: Statuses[] = [
      ["user", "D", "private", "ongoing", "event", "active", "punctual"],
      ["admin", "S", "public", "done", "class", "archive", "overdue"],
    ];

    const list = samples.map((x) => joinStrings(x));
    expect(list).toMatchSnapshot();
  });
});
