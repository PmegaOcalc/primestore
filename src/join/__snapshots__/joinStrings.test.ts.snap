// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinStrings function renders all possible strings correctly 1`] = `
[
  [
    "user",
    "user_D",
    "user_D_private",
    "user_D_private_ongoing",
    "user_D_private_ongoing_event",
    "user_D_private_ongoing_event_active",
    "user_D_private_ongoing_event_active_punctual",
    "D",
    "D_private",
    "D_private_ongoing",
    "D_private_ongoing_event",
    "D_private_ongoing_event_active",
    "D_private_ongoing_event_active_punctual",
    "private",
    "private_ongoing",
    "private_ongoing_event",
    "private_ongoing_event_active",
    "private_ongoing_event_active_punctual",
    "ongoing",
    "ongoing_event",
    "ongoing_event_active",
    "ongoing_event_active_punctual",
    "event",
    "event_active",
    "event_active_punctual",
    "active",
    "active_punctual",
    "punctual",
  ],
  [
    "admin",
    "admin_S",
    "admin_S_public",
    "admin_S_public_done",
    "admin_S_public_done_class",
    "admin_S_public_done_class_archive",
    "admin_S_public_done_class_archive_overdue",
    "S",
    "S_public",
    "S_public_done",
    "S_public_done_class",
    "S_public_done_class_archive",
    "S_public_done_class_archive_overdue",
    "public",
    "public_done",
    "public_done_class",
    "public_done_class_archive",
    "public_done_class_archive_overdue",
    "done",
    "done_class",
    "done_class_archive",
    "done_class_archive_overdue",
    "class",
    "class_archive",
    "class_archive_overdue",
    "archive",
    "archive_overdue",
    "overdue",
  ],
]
`;
